### PR TITLE
WFLY-9563 Integrate KUBE_PING discovery protocol

### PR DESCRIPTION
**Pull request by rhusar**
Jira
https://issues.jboss.org/browse/WFLY-9563

Proposal
https://github.com/wildfly/wildfly-proposals/pull/10

### CODE REVIEW

File: feature-pack/src/main/resources/modules/system/layers/base/org/jgroups/kubernetes/main/module.xml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  ~ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+  ~ Lesser General Public License for more details.
+  ~
+  ~ You should have received a copy of the GNU Lesser General Public
+  ~ License along with this software; if not, write to the Free
+  ~ Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+  ~ 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+  -->
+
+<module xmlns="urn:jboss:module:1.7" name="org.jgroups.kubernetes">
```
Comment (LittleJohnII) [inline feature-pack/src/main/resources/modules/system/layers/base/org/jgroups/kubernetes/main/module.xml]: My 2 cents: unless there's a reason for making this public, I suggest making this module unsupported, like we do with azure ping.

**Review by rjanik114 — APPROVED** [review]
LGTM, one comment that might or might not be relevant.

File: feature-pack/src/main/resources/modules/system/layers/base/org/jgroups/kubernetes/main/module.xml (hunk 1 above)
Comment (rhusar) [reply feature-pack/src/main/resources/modules/system/layers/base/org/jgroups/kubernetes/main/module.xml]: No, we are not making `org.jgroups.azure` unsupported, see: https://github.com/wildfly/wildfly/blob/master/feature-pack/src/main/resources/modules/system/layers/base/org/jgroups/azure/main/module.xml#L25

Comment (LittleJohnII) [reply feature-pack/src/main/resources/modules/system/layers/base/org/jgroups/kubernetes/main/module.xml]: Oh, ok, that's what I get for checking EAP not WildFly. Apparently, it's only unsupported there: https://github.com/jbossas/jboss-eap7/blob/7.1.x/feature-pack/src/main/resources/modules/system/layers/base/org/jgroups/azure/main/module.xml#L28

Nothing then.

**Comment by stuartwdouglas** [comment]
Why is this marked [hold], was it just because it was for 13.x or is there some other reason?

**Comment by stuartwdouglas** [comment]
@rhusar please ping me when this is no longer a hold and I will remove the label.

**Comment by kabir** [comment]
@stuartwdouglas @rhusar I guess it is hold because WFLY-9563 is a feature request and it is missing some stuff to get merged. Adding Feature and removing the hold label

File: feature-pack/pom.xml [hunk 2]
```diff
@@ -1431,6 +1431,15 @@
             <artifactId>jgroups-azure</artifactId>
         </dependency>
 
+        <dependency>
+            <groupId>org.jgroups.kubernetes</groupId>
+            <artifactId>jgroups-kubernetes</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>net.oauth.core</groupId>
```
Comment (slaskawi) [inline feature-pack/pom.xml]: With the latest version this is no longer necessary. See https://github.com/jgroups-extras/jgroups-kubernetes/commit/08e6455290cae780ef26e07aa593e849bcdf2f2f

**Review by slaskawi — CHANGES_REQUESTED** [review]
Apart from dropping oauth library, LGTM.

**Comment by rhusar** [comment]
Great, thank you @slaskawi ! 1.0.6.Final was not out when I initially opened the PR, so removed the dependency now. What an Irony... we setup jgroups-extras so we can bring in dependencies.

**Comment by rhusar** [comment]
@kabir I have held the PR to coordinate with other teams.
@wildfly/mergers-full This should be ready.

**Comment by rhusar** [comment]
@LittleJohnII  Any QE update on this?

**Comment by LittleJohnII** [comment]
@rhusar not yet, no. I still haven't pretested this. I'll let you know when that's done. Expect tomorrow or monday, latest.

**Comment by LittleJohnII** [comment]
@rhusar I'm working on the precheck, but running into issues (not with this PR, that looks good so far). I'll let you know again tomorrow.

File: docs/src/main/asciidoc/_high-availability/subsystem-support/Kubernetes.adoc [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+Discovery is therefore done by asking Kubernetes for a list of IP addresses of all cluster nodes.
+Combined with `bind_port` / `port_range`, the protocol will then send a discovery request to all instances and wait for the responses.
+
+To configure an existing profile with `KUBE_PING` use the following CLI batch replacing with appropriate stack:
+
+[source]
+----
+batch
+/subsystem=jgroups/stack=tcp/protocol=MPING:remove()
+/subsystem=jgroups/stack=tcp/protocol=kubernetes.KUBE_PING:add(add-index=0,properties={namespace="production",labels="cluster=nyc")}
```
Comment (LittleJohnII) [inline docs/src/main/asciidoc/_high-availability/subsystem-support/Kubernetes.adoc]: The `}` and `)` at the end of the line have been swapped here.

Comment (rhusar) [reply docs/src/main/asciidoc/_high-availability/subsystem-support/Kubernetes.adoc]: Fixed, thanks. Also expanded the document a little while at it.

**Comment by LittleJohnII** [comment]
@rhusar @kabir The precheck is complete. I run into some issues, none blocking. Most of them don't even relate to KUBE_PING and are likely related to my environment. I'll describe everything in the planning issue. The one thing that's worth mentioning here is that KUBE_PING logs two warnings which are redundant and IMHO shouldn't be logged by the server:

```
12:26:03,871 WARN  [org.jgroups.protocols.kubernetes.KUBE_PING] (ServerService Thread Pool -- 68) OPENSHIFT_KUBE_PING_NAMESPACE is deprecated, please remove it and use KUBERNETES_NAMESPACE instead
12:26:03,871 WARN  [org.jgroups.protocols.kubernetes.KUBE_PING] (ServerService Thread Pool -- 68) OPENSHIFT_KUBE_PING_LABELS is deprecated, please remove it and use KUBERNETES_LABELS instead
```

I did not specify any environment variables since the namespace and labels are a part of the server configuration. This is due to the `if` here: https://github.com/jgroups-extras/jgroups-kubernetes/blob/master/src/main/java/org/jgroups/protocols/kubernetes/KUBE_PING.java#L176
This should IMHO look like:
```
if (propertyDefined && deprecatedDefined)
    log.warn("specify one");
else if (deprecatedDefined)
    log.warn("deprecated...");
}
```
But this is no big deal. I'll file a JGRP issue for this.

What I tested is a scale-up/scale-down scenario with 4 nodes on a local kubernetes instance via Minikube as well as filtering based on namespaces and labels - all of that behaves as expected.

In other words, I think we can move forward with this if you want. I still need to automate my tests, but I'll track that on the planning issue.
